featuretests/application_config_test.go

```diff
@@ -0,0 +1,223 @@
+			"floatoverwrite":   2.1,
+			"booleanoverwrite": false,
+		},
+	})
+	c.Assert(err, jc.ErrorIsNil)
+
+	unit, err := app.AddUnit()
+	c.Assert(err, jc.ErrorIsNil)
+	unit.SetCharmURL(s.charm.URL())
+
+	password, err := utils.RandomPassword()
+	c.Assert(err, jc.ErrorIsNil)
+	err = unit.SetPassword(password)
+	c.Assert(err, jc.ErrorIsNil)
+
+	st := s.OpenAPIAs(c, unit.Tag(), password)
+	uniteer, err := st.Uniter()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(uniteer, gc.NotNil)
+
+	s.apiUnit, err = uniteer.Unit(unit.Tag().(names.UnitTag))
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Ensure both outputs have all charm config keys
+	s.settingKeys = set.NewStrings()
+	for k, _ := range s.charm.Config().Options {
+		s.settingKeys.Add(k)
+	}
+}
+
+func (s *ApplicationConfigSuite) configCommandOutput(c *gc.C, args ...string) string {
+	context, err := testing.RunCommand(c, application.NewConfigCommand(), args...)
+	c.Assert(err, jc.ErrorIsNil)
+	return testing.Stdout(context)
+}
+
+func (s *ApplicationConfigSuite) getHookOutput(c *gc.C) charm.Settings {
+	settings, err := s.apiUnit.ConfigSettings()
+	c.Assert(err, jc.ErrorIsNil)
+	return settings
+}
+
+// The primary of objective of this test is to ensure
+// that both 'juju get' as well as unit in a hook context, uniter.unit, agree
+// on all returned settings and values.
+// These implementations are separate and cannot be re-factored. However,
+// since the logic and expected output is equivalent, these should be modified in sync.
+func (s *ApplicationConfigSuite) TestConfigAndConfigGetReturnAllCharmSettings(c *gc.C) {
+	// initial deploy with custom settings
+	s.assertSameConfigOutput(c, initialConfig)
+
+	// use 'juju config foo=' to change values
+	s.configCommandOutput(c, s.appName,
+		"booleandefault=false",
+		"booleannodefault=true",
+		"booleanoverwrite=true", //charm default
+		"floatdefault=7.2",
+		"floatnodefault=10.2",
+		"floatoverwrite=11.1", //charm default
+		"intdefault=22",
+		"intnodefault=11",
+		"intoverwrite=111", //charm default
+		"strdefault=not",
+		"strnodefault=maybe",
+		"stroverwrite=me",
+	)
+	s.assertSameConfigOutput(c, updatedConfig)
+
+	// 'juju config --reset' to reset settings to charm default
+	s.configCommandOutput(c, s.appName, "--reset",
+		"booleandefault,booleannodefault,booleanoverwrite,floatdefault,"+
+			"floatnodefault,floatoverwrite,intdefault,intnodefault,intoverwrite,"+
+			"strdefault,strnodefault,stroverwrite")
+	s.assertSameConfigOutput(c, resetConfig)
+}
+
+func (s *ApplicationConfigSuite) assertSameConfigOutput(c *gc.C, expectedValues settingsMap) {
+	s.assertJujuConfigOutput(c, s.configCommandOutput(c, s.appName), expectedValues)
+	s.assertHookOutput(c, s.getHookOutput(c), expectedValues)
+}
+
+func (s *ApplicationConfigSuite) assertHookOutput(c *gc.C, obtained charm.Settings, expected settingsMap) {
+	c.Assert(len(obtained), gc.Equals, len(expected))
+	c.Assert(len(obtained), gc.Equals, len(s.settingKeys))
+	for name, aSetting := range expected {
```

Comment:
I would have thought for 'expected' values which are 'nil', that the hook output would just omit them, rather than transmitting 'nil' as a value. Or is that part of what you changed?

eg, I would have thought there would be a pass where 'expected' has 'nil' values filtered out.